---
title: Slamdunk paper
description: SLAMseq analysis using Slamdunk for nucleotide-conversion sequencing datasets
date: 2019-06-28 13:42
image:
  path: /assets/images/categories/logo_slamdunk_rgb_72.jpg
categories:
  - Pipelines
tags:
- SLAMseq
- Slamdunk
- Containers
- Bioconda
- PyPI
- Docker
toc: true
toc_sticky: true
author_profile: false
---

For the past couple of years I was involved in the development of [SLAMseq](http://doi.org/10.1038/nmeth.4435), a sequencing technology for time-resolved measurement of newly synthesized and existing RNA in cultured cells. Originally developed by the lab of [Stefan Ameres](https://www.imba.oeaw.ac.at/research/stefan-ameres/), the lab of my boss [Johannes Zuber](https://www.imp.ac.at/groups/johannes-zuber/) extended the approach with pharmacological and chemical-genetic perturbations in order to identify direct transcriptional targets of any gene or pathway ([Muhar et al, Science 2018](http://doi.org/10.1126/science.aao2793)).

Processing and interpreting this data required novel analysis methods, so I was given the opportunity to team up with a good friend of mine - [Philipp Rescheneder](https://github.com/philres) - to develop [Slamdunk](https://t-neumann.github.io/slamdunk/) which we recently published in [BMC Bioinformatics](http://doi.org/10.1186/s12859-019-2849-7) and is generally applicable to any nucleotide-conversion containing dataset.

This post will quickly highlight the main functionality, findings and features.

## Slamdunk workflow

<img src="{{ site.url }}{{ site.baseurl }}/assets/images/posts/Slamdunk/slamdunk_outline.png" alt="Slamdunk outline">

Slamdunk differs from naive read processing in 4 ways:

* It maps with a nucleotide-conversion aware scoring scheme since in the example of SLAMseq data, T>C mismatches are expected and identify reads from labelled transcripts
* Since QuantSeq processes smaller, more repetitive regions of transcripts - namely the 3' ends - Slamdunk cannot simply discard all multimappers, but utilizes a strategy to recover them
* Genuine T>C SNPs would contribute greatly to false-positive conversion-quantifications and have to be excluded during the quantification step
* Depending on coverage and T-content in the 3' end, observing T>C reads will have a different likelihood which has to be corrected for during conversion quantification

## Availability

Slamdunk is available from several platforms:

* [Github](http://doi.org/10.1126/science.aao2793) @ icon-github
